src/datasources/accounts/counterfactual-safes/counterfactual-safes.datasource.ts

```diff
@@ -0,0 +1,145 @@
+import { IConfigurationService } from '@/config/configuration.service.interface';
+import { CacheRouter } from '@/datasources/cache/cache.router';
+import {
+  CacheService,
+  ICacheService,
+} from '@/datasources/cache/cache.service.interface';
+import { getFromCacheOrExecuteAndCache } from '@/datasources/db/utils';
+import { CounterfactualSafe } from '@/domain/accounts/counterfactual-safes/entities/counterfactual-safe.entity';
+import { CreateCounterfactualSafeDto } from '@/domain/accounts/counterfactual-safes/entities/create-counterfactual-safe.dto.entity';
+import { Account } from '@/domain/accounts/entities/account.entity';
+import { ICounterfactualSafesDatasource } from '@/domain/interfaces/counterfactual-safes.datasource.interface';
+import { ILoggingService, LoggingService } from '@/logging/logging.interface';
+import { Inject, Injectable, NotFoundException } from '@nestjs/common';
+import postgres from 'postgres';
+
+@Injectable()
+export class CounterfactualSafesDatasource
+  implements ICounterfactualSafesDatasource
+{
+  private readonly defaultExpirationTimeInSeconds: number;
+
+  constructor(
+    @Inject(CacheService) private readonly cacheService: ICacheService,
+    @Inject('DB_INSTANCE') private readonly sql: postgres.Sql,
+    @Inject(LoggingService) private readonly loggingService: ILoggingService,
+    @Inject(IConfigurationService)
+    private readonly configurationService: IConfigurationService,
+  ) {
+    this.defaultExpirationTimeInSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.default',
+      );
+  }
+
+  // TODO: the repository calling this function should:
+  // - check the AccountDataSettings to see if counterfactual-safes is enabled.
+  // - check the AccountDataType to see if it's active.
```

Comment:
I've just left this TODO here for reference when implementing the Repository that will use this datasource.